Replace iodb with leveldb

knizhnik opened the pull request.

oskin1 commented on src/main/scala/org/ergoplatform/nodeView/state/DigestState.scala:
```diff
@@ -159,31 +160,34 @@ object DigestState extends ScorexLogging with ScorexEncoding {
+          } else {
+            val inVersion = store.lastVersionID.map(w => bytesToVersion(w.data)).getOrElse(version)
+            new DigestState(inVersion, rootHash, store, constants.settings)
+              .update(version, rootHash, context).get //sync store
```
Try to avoid using `.get` at all, even inside a `Try`.

knizhnik replied:
Опять таки код не мой и как эту ситуацию правильно обрабатывать я не очень понимаю.

knizhnik commented on avldb/src/main/scala/scorex/crypto/authds/avltree/batch/NodeParameters.scala:
```diff
@@ -0,0 +1,5 @@
+package scorex.crypto.authds.avltree.batch
+
+//todo: support arbitrary-size values
+@specialized
```
Это не мой код - так было. Могу удалить.

knizhnik commented on avldb/src/main/scala/scorex/db/LDBFactory.scala:
```diff
@@ -0,0 +1,122 @@
+    val pairs = loaders.view
+      .zip(factories)
+      .flatMap { case (loader, factoryName) =>
+        Try(loader.loadClass(factoryName).getConstructor().newInstance().asInstanceOf[DBFactory]).toOption
```
Перенёс фикс.

aslesarenko commented on avldb/src/main/scala/scorex/db/LDBFactory.scala:
```diff
@@ -0,0 +1,128 @@
+  val lock = new ReentrantReadWriteLock()
+  val map = new mutable.HashMap[File, RegisteredDB]
+
+  case class RegisteredDB(val impl:DB, val path: File) extends DB {
```
Please document the role and purpose of this classes.

knizhnik replied:
Done

aslesarenko commented on avldb/src/main/scala/scorex/db/LDBFactory.scala:
```diff
@@ -0,0 +1,128 @@
+
+case class StoreRegistry(val factory : DBFactory) extends DBFactory {
+
+  val lock = new ReentrantReadWriteLock()
```
Please comment in the code on the motivation to use this specific type of lock.

knizhnik replied:
Multithreaded application may try to create several database at the same time.
To avoid race condition in StoreRegistry we need to synchronize access to this class.
As far as I understand, ReentrantReadWriteLock is standard way of synchronizing access to the class (replacement of Java method "synchronized" qualifier).

aslesarenko commented on avldb/src/main/scala/scorex/db/LDBFactory.scala:
```diff
@@ -0,0 +1,128 @@
+
+  lazy val factory: DBFactory = {
+    val loaders = List(ClassLoader.getSystemClassLoader, this.getClass.getClassLoader)
+    val factories = List(nativeFactory, javaFactory)
```
@kushti @oskin1 I suggest to remove `javaFactory` from this list.

knizhnik replied:
As far as I remember, the presence of experimental Java implementation of leveldb was motivated by the fact that not all target platforms have leveldb installed.
If we are not going to include leveldb binaries in our distributive, then we should allow to use Java leveldb port in this case.

oskin1 replied:
We are already including leveldb binaries into ergo jar. 
But java version should be left here as fallback implementation. It already helped guys who ran the node on rare platform so there weren’t native leveldb binaries available for it.

aslesarenko replied:
@oskin1 ok, but we know Java version is unstable, so it has its own risks. Then we need to add warning messages to log with big exclamation marks and disclaimer, that it is not supported and experimental feature.

aslesarenko commented on avldb/src/main/scala/scorex/db/LDBVersionedStore.scala:
```diff
@@ -0,0 +1,316 @@
+
+
+
+class LDBVersionedStore(val dir: File, val keepVersions: Int = 0) extends Store {
```
Add documentation for this class

knizhnik replied:
Done

aslesarenko commented:
@knizhnik Please describe where this code come from, what is new, and what is copied from some other place? And also some motivation. Should I review all the code, or only new parts? /cc @kushti

aslesarenko commented:
@knizhnik also, when you adding the code, you becoming its owner. If most of the code is copied from somewhere, it is better to split the PR in two parts: 1) add all code 2) add your modifications.
For example #944 was pure copy of the code from external repository, so no review was required other that file-by-file comparisons of directories.

knizhnik commented on avldb/benchmarks/src/main/scala/scorex/crypto/authds/benchmarks/Helper.scala:
```diff
@@ -0,0 +1,91 @@
+
+    val inserts = (r.head until r.head + insertsCount).map { i =>
+      val key = ADKey @@ new Array[Byte](kl)
+      val k = Longs.toByteArray(i) ++ Longs.toByteArray(System.currentTimeMillis)
```
Key size in IODB is fixed, This is key is generated in such strange way. 
There is no such limitation in levelDB.

knizhnik commented on avldb/build.sbt:
```diff
@@ -0,0 +1,42 @@
+)
+
+testOptions in Test := Seq(Tests.Filter(t => !t.matches(".*Benchmark$")))
+javaOptions in run += "-Xmx12G"
```
Such values was needed only to pass all configurations of AVLTreeBatchPerformance (the largest one is  StateWith32000000)

knizhnik commented:
Actually I have implemented only one file
ergo/avldb/src/main/scala/scorex/db/LDBVersionedStore.scala,
added avldb as subproject of ergo, moved LDBFactory from ergo to avldb
and made some small fixes in avldb and ergo projects (deletion of database files, specifying hostory size,...).

knizhnik commented on avldb/build.sbt:
```diff
@@ -0,0 +1,42 @@
+import sbt.Keys.testFrameworks
+
+name := "avldb"
```
I did it in the same way as ergo-walet was integrated in ergo project.
I am not expert in Scala and especially SBT, so may be it can done in some other way.

aslesarenko commented:
@knizhnik I would like to see a diff of the changes made in the original code. The PR shows all files as new even if they were not changed. This means I cannot evaluate what was changed, and cannot estimate the impact of those changes.

knizhnik commented:
In AVLDB I have made only changes related with replacing deteriorated class, i.e.

< import scorex.crypto.encode.Base58
---
> import scorex.util.encode.Base58

I can send you doff file, but actually there is nothing interesting in it: only few such refactoring changes, mostly autosuggested by Idea.

And changes done in ergo project itself are in commits 0e49da007ddf54, d25254d205, a230bbfbc8

aslesarenko commented:
Replaced by #969 and #968
